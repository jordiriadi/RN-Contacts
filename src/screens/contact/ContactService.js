import Sql from '../../Sql/client';
import Contact from './types/Contact';

export function getContacts() {
  const extractPayload = (payload) => payload.data.map(Contact);

  return Sql.query({}, { method: 'GET' }).then(extractPayload);
}

export function addContact(formValues) {

  return Sql.query({}, { method: 'POST', query: formValues});
}

export function updateContact(id, _formValues) {
  const formValues = {
    firstName: _formValues.firstName,
    lastName: _formValues.lastName,
    age: _formValues.age
  };
  const stringifyParams = `/${id}`;

  return Sql.query({}, { method: 'PUT', query: formValues, stringifyParams})
}

export function deleteContact(id) {
  const stringifyParams = `/${id}`;

  return Sql.query({}, { method: 'DELETE', stringifyParams});
}

export default {
  getContacts,
  addContact,
  updateContact,
  deleteContact
}