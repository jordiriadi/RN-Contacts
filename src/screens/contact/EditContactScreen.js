import React from 'react';
import { StyleSheet, View } from 'react-native';

import ModifyContactForm from './ModifyContactForm';
import FlashMessage from '../../components/FlashMessage';
import ContactService from './ContactService';
import Styles from '../../utils/Styles';

class EditContactScreen extends React.Component {
  constructor(props){
    super(props);

    this.state = {
      loading: false,
      error: null,
    }
  }

  goToDetailScreen = (contact) => this.props.navigation.navigate('ContactDetailScreen', {contact});

  handleUpdateContact = (formValues) => {
    console.log('update contact ...');
    const {id} = formValues;
    const updateContact = ContactService.updateContact(id, formValues);

    const startLoading = () => this.setState({ loading: true });
    const endLoading = () => this.setState({ loading: false });
    
    const handleSuccess = (res) => {
      endLoading();
      this.goToDetailScreen(res.data);
    }
    const handleFail = (err) => {
      this.setState({
        error: err,
        loading: false,
      });
      console.log(err);
    }


    startLoading();
    updateContact.then(handleSuccess, handleFail).catch(handleFail);


    return updateContact;
  }

  handleResetError = () => this.setState({ error: null });
  
  handleCancel = () => this.props.navigation.goBack();

  render() {
    const { error } = this.state;
    const { contact } = this.props.navigation.state.params || {};

    return (
      <View style={styles.container}>
        {error && (
          <FlashMessage
            text={error}
            onTimeout={this.handleResetError}
          />
        )}
        <ModifyContactForm
          modify={"update"} 
          value={contact}
          onSave={this.handleUpdateContact} 
          onCancel={this.handleCancel}
          isLoading={this.state.loading} 
          />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingTop: Styles.DefaultPadding*8,
    backgroundColor: Styles.PrimaryColor,
  }
})


export default EditContactScreen;