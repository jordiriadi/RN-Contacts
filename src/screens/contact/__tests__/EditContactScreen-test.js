import 'react-native';
import React from 'react';
import {shallow} from 'enzyme';

import EditContactScreen from '../EditContactScreen';
import ContactService from '../ContactService';

const mockData = {
  "age": "33",
  "firstName": "LukeTheSecond",
  "id": "b3abd640-b22g-11e8-b02f-cbfa15db428b",
  "lastName": "Skywalker", 
  "photo": "N/A"
}

const mockProps = {
  navigation: {
    state: {
      params: {
        contact: mockData,
      }
    },
    setParams: jest.fn(),
    push: jest.fn(),
    navigate: jest.fn(),
    goBack: jest.fn(),
  }
}

ContactService.updateContact = jest.fn(() => Promise.resolve(mockData));

describe('Edit Contact Screen', () => {  
  it('renders correctly', () => {
  
    const wrapper = shallow(<EditContactScreen {...mockProps}/>);

    expect(wrapper).toMatchSnapshot();
  });

  it('can trigger handle edit contact', () => {
    const component = shallow(<EditContactScreen {...mockProps}/>);

    expect(component.instance().handleUpdateContact(mockData));
  });
})