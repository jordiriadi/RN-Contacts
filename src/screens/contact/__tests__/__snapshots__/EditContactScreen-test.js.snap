// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Contact Screen renders correctly 1`] = `ShallowWrapper {}`;
